Add types for DjangoFilterBackend

last-partizan opened the pull request:
After recent changes to pyright, we need to add proper types for django-filters for it to pass type-checking.

**Required** changes is only `filter_queryset` types, but i have added other as well.

~~This is draft, and it still not working as expected, but i will update this and djangoresframework-types after some discussion in pyright repo, on how to do this better.~~

~~Main problem is that `djangorestframework-types` have included MongoQuerySet in it's types, and i don't want to drag it here. Maybe it's better to replace it with some sort protocol there.~~

last-partizan commented:
Update: we removed MongoQuerySet support from djangorestframework types and now this is ready to merge. It's working as expected with latest pyright.

@carltongibson can we merget this? or maybe you have some suggestions for improvements?

carltongibson commented:
Hi. TBH I'm not keen on the inline annotations here. It seems like a lot of noise. 🤔 

Could you use a stub file? 

(Didn't I read they changed the `Optional` syntax?)

last-partizan commented:
Yes, new Optional and Union syntax available starting from Python 3.10, but probably we can use it with 'from __future__` import.

I've cleaned up it a bit and updated to use new syntax.

I theory, yes, i can use type stub - but it would require me to maintain separate package which will break if anything changed here.

It may seem as noise at first, and typing in python has some rough edges, but developer expirience is improved when we have types. I'm started adding types to my projects a few years ago, and it helped me to catch a few bugs.

That's especially useful when types are added to projects, not as separate stub files.

carltongibson commented:
Nothing is going to change here 😅

last-partizan commented:
Oh, okay.

carltongibson commented:
Hi @last-partizan -- good hustle 😜

I'm not totally opposed to the idea of the stub file... 

My concerns would be: 

1. I'd have to maintain it. 
2. I don't really see the point of types without type checking, so there's (massive) scope creep here just getting out the gate. 

I'd lean towards you just adding a stub file in your own project to begin, but if you want to open an exploratory PR here, I'm happy to look at that (and I would do so with an open mind.) 

Hope that makes sense.

last-partizan commented:
Yeah, maintaining separate type files is a fuss, that's why i like types in a code.

And maintaining my own stub files is yet bigger fuss.

I'd probably just use my fork with types, it would be easier to maintain in case of changes.

Have a nice day!

last-partizan commented:
Hello again :)

Turns out, maintaining a fork is more troublesome than I expected. But I got it updated, and it works again.

Let's try another attempt on merging this, as inline types, not as separate `.pyi`.

First, I'll try to address your concerns: 

1. Inline types doesn't need any maintenance - they're just exists. No additional work required. When you add new variable - you add type. That's all. If you have any troubles with types, just ping me - I'm here to help.
2. Main point of adding types here - remove those annoying pyright errors for everyone using this package together with pyright. (And potentially other type checkers). You don't need to add types to entire package or add type-checking to CI (but, if you want to have type checking in CI, I can create PR).

If you're concerned about "adding noise", I can make this one line change (+few lines of imports). `djangorestframework-types` requires only `filter_queryset` method to be typed, so I can remove everything else.

What do you think? I'll create another PR, instead of this outdated one, if you're open to adding this.

carltongibson commented on django_filters/rest_framework/backends.py:
```diff
@@ -1,30 +1,40 @@
         return filterset_class(**kwargs)
 
-    def get_filterset_class(self, view, queryset=None):
+    def get_filterset_class(self, view: GenericAPIView, queryset: QuerySet[Any] | None = None) -> type[filterset.FilterSet] | None:
```
If I'm honest, I just don't want this in my code. `queryset: QuerySet[Any]` — like really? 

(sigh)

last-partizan replied:
Well, it's the best we have for now, and it's very useful.

Actually, there should be `QuerySet[_MT]`, where `_MT = TypeVar("_MT", bound=Model)`, and it allows type-checker and LSP server to know that when you access `queryset.model._meta` - it's model options, and your editor can suggest all relevant attributes when you need it.

carltongibson replied:
@last-partizan Yes, I understand how it works. I just don't like looking at it, or writing it.

carltongibson commented:
@last-partizan I'm a bit sceptical to be honest. 

(IDK)

last-partizan commented:
This time I was able to fix issue at the core - I found issue with `djangorestframework-types`, which prevented a protocol from matching this untyped class.

Now I can have my sweet type checking, and you can have your clean code without types. At least for now :)

Sorry for bothering you, and thanks for this amazing project!

carltongibson commented:
@last-partizan I'm not totally anti here. I am on the negative side still though.
